src/lowering/desugar.ml

```diff
@@ -386,32 +407,39 @@ and build_actor at ts self_id es obj_typ =
   let meta =
     I.{ candid = candid;
         sig_ = T.string_of_stab_sig sig_} in
-  let (interface_d, interface_f) = export_interface candid.I.service in
-  I.ActorE (interface_d @ ds', interface_f @ fs,
+  let interface_d, interface_f = export_interface candid.I.service in
+  let with_stable_vars wrap =
+    let vs = fresh_vars "v" (List.map (fun f -> f.T.typ) fields) in
+    blockE
+      ((match call_system_func_opt "preupgrade" es with
+        | Some call -> [ expD (primE (I.OtherPrim "⏰rts_perform_gc⏰") [])
+                       ; expD call]
+        | None -> []) @
+         [letP (seqP (List.map varP vs)) (* dereference any mutable vars, option 'em all *)
```

Comment:
Note this will only do a GC if we have a preupgrade system method, not for every size computation.

Reply:
That is intended. Why lose time when size computation doesn't care about fragmentation, and the heap (potentially damaged) will be jettisoned anyway? We might find an argument for doing a GC _after_ having the stable variables ready for serialisation, but that is another PR.

Reply:
Oh, you're right. I was assuming we might need to allocate but we don't.